File: library/src/main/java/com/dylanc/retrofit/helper/transformer/ThreadTransformer.java
package com.dylanc.retrofit.helper.transformer;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.ObservableTransformer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * @author Dylan Cai
 * @since 2019/7/11
 */
public class ThreadTransformer {

  public static <T> ObservableTransformer<T, T> main() {
    return new ObservableTransformer<T, T>() {
      @Override
      public ObservableSource<T> apply(Observable<T> upstream) {
        return upstream.subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread());
      }
    };
  }
}

File: library/src/main/java/com/dylanc/retrofit/helper/transformer/ThreadTransformer.kt
package com.dylanc.retrofit.helper.transformer

import io.reactivex.Observable
import io.reactivex.ObservableTransformer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import me.jessyan.progressmanager.ProgressListener
import okhttp3.ResponseBody

/**
 * @author Dylan Cai
 * @since 2019/7/13
 */

fun <T> Observable<T>.io2mainThread(): Observable<T> = compose(ThreadTransformer.io2main())

object ThreadTransformer {
  @JvmStatic
  fun <T> io2main(): ObservableTransformer<T, T> {
    return ObservableTransformer { upstream ->
      upstream.subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
    }
  }
}
